pkg/kv/kvserver/replica_raftlog.go

```diff
@@ -0,0 +1,275 @@
+func (r *replicaLogStorage) termLocked(i kvpb.RaftIndex) (kvpb.RaftTerm, error) {
+	// TODO(pav-kv): make it possible to read with only raftMu held.
+	r.mu.AssertHeld()
+	// TODO(nvanbenschoten): should we set r.mu.lastTermNotDurable when
+	//   r.mu.lastIndexNotDurable == i && r.mu.lastTermNotDurable == invalidLastTerm?
+	// TODO(pav-kv): we should rather always remember the last entry term, and
+	// remove invalidLastTerm special case.
+	if r.mu.lastIndexNotDurable == i && r.mu.lastTermNotDurable != invalidLastTerm {
+		return r.mu.lastTermNotDurable, nil
+	}
+	return logstore.LoadTerm(r.AnnotateCtx(context.TODO()),
+		r.mu.stateLoader.StateLoader, r.store.TODOEngine(), r.RangeID,
+		r.store.raftEntryCache, i,
+	)
+}
+
+// raftTermLocked implements the Term() call.
+func (r *Replica) raftTermLocked(i kvpb.RaftIndex) (kvpb.RaftTerm, error) {
+	return (*replicaLogStorage)(r).termLocked(i)
+}
+
+// GetTerm returns the term of the entry at the given index in the raft log.
+// Requires that r.mu is not held.
+func (r *Replica) GetTerm(i kvpb.RaftIndex) (kvpb.RaftTerm, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.raftTermLocked(i)
+}
+
+// LastIndex implements the raft.LogStorage interface.
+// Requires that r.mu is held for reading.
+func (r *replicaLogStorage) LastIndex() uint64 {
+	return uint64((*Replica)(r).raftLastIndexRLocked())
+}
+
+// raftLastIndexRLocked implements the LastIndex() call.
+func (r *Replica) raftLastIndexRLocked() kvpb.RaftIndex {
+	return r.mu.lastIndexNotDurable
+}
+
+// GetLastIndex returns the index of the last entry in the raft log.
+// Requires that r.mu is not held.
+func (r *Replica) GetLastIndex() kvpb.RaftIndex {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.raftLastIndexRLocked()
+}
+
+// FirstIndex implements the raft.LogStorage interface.
+// Requires that r.mu is held for reading.
+func (r *replicaLogStorage) FirstIndex() uint64 {
+	return uint64((*Replica)(r).raftFirstIndexRLocked())
+}
+
+// raftFirstIndexRLocked implements the FirstIndex() call.
+func (r *Replica) raftFirstIndexRLocked() kvpb.RaftIndex {
+	// TruncatedState is guaranteed to be non-nil.
+	return r.mu.state.TruncatedState.Index + 1
+}
+
+// GetFirstIndex returns the index of the first entry in the raft log.
+// Requires that r.mu is not held.
+func (r *Replica) GetFirstIndex() kvpb.RaftIndex {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.raftFirstIndexRLocked()
+}
+
+// LogSnapshot returns an immutable point-in-time snapshot of the log storage.
+// Requires that r.raftMu is held for writing, and r.mu for reading.
+func (r *replicaLogStorage) LogSnapshot() raft.LogStorageSnapshot {
+	r.raftMu.AssertHeld()
+	r.mu.AssertRHeld()
+	return (*replicaLogSnap)(r)
+}
+
+// replicaLogSnap implements the raft.LogStorageSnapshot interface.
+//
+// The type implements a limited version of a raft log storage snapshot, without
+// needing a storage engine snapshot. It relies on the fact that all raft writes
```

Comment:
Related to my comment below - since `raftMu` being held throughout is what _makes_ this a snapshot, should we rename this to `replicaRaftMuLogSnap`? I think that drives home nicely the crucial bit of information which I worry will be lost on readers or drive confusion, which is that this kind of snapshot is essentially an "optimization" over the "obvious" implementation which would have us grab a consistent iterator upon construction: instead we make use of the fact that `raftMu` prevents mutation and so any "actual" snapshotting can be elided. (We should talk about it like that too in the type comments etc, btw, unless I'm wrong).

Reply:
The part that I still find tricky is that we're then "hiding" this raftMu-edness by going through these interfaces. We have to somehow make it obvious that raft has this interface method to get a snapshot, but that we know so much about the circumstances in which this will be called (raftMu is held throughout) that we can plop our "shim" implementation in that doesn't actually need real snapshotting.

I think there is something worth cleaning up here. The raft interface currently is obfuscating what is going on, rather than help explain it and I would prefer to address this or at least understand how we could. I think this problem is similar to the one we have with all of the existing methods on `raft.Storage` so there is a good payoff beyond just raft storage snapshots.

The core problem seems to be that currently, `RawNode` obtains the snapshot and we rely on our knowledge that it only calls this particular method in the `LogStorage` interface in methods that we call with certain mutexes held. If we always passed in the storage with each call that needs it, this problem would go away or at least become significantly more transparent.

For example, instead of `rawNode.Ready()` calling `rawNode.storage.Entries()` which in turn is our `*Replica`-level implementation that assumes certain locks are held by the caller of `Ready`, we would have something like

```go
r.raftMu.Lock()
r.mu.Lock()

rd := rawNode.Ready(raftMuAndMuHeldRaftStorage(*Replica))
// ...

r.mu.Unlock()
r.raftMu.Unlock()

```

It also seems like a somewhat moderate clean-up that is mostly mechanical, but still has the potential to make this whole locking business vastly more understandable, perhaps to the point of it being "obviously correct".

Reply:
I agree that making things more explicit / less callback-ey would be great, I'll explore in follow-up after getting something working for RACv2.

There is a slight benefit in the current approach, in that there is one `LogStorage` object inside raft, so we have a guarantee that all snapshots are for the same object. By passing the snapshot into `Ready`, we would "denormalize" this (raft has `LogStorage` internally, but also receives snapshots from the caller - and they must match what it has internally). So if we flip the API to accept storage from outside, we should do it all the way: raft would have no storage to read from except when we tell it. Or bring it to next level, and always call log storage methods on our end explicitly, and raft would only instruct us which log slices it's interested in.

The heavy users of the log storage are `MsgApp`, local log appends, local state machine application. Eventually we want to flip control and pace all of them (via the enhanced `Ready` API or 3 parallel smaller ones). So flipping the storage interface ownership/passing fits.

After this, `LogStorage` can be removed from `RawNode` because it won't have readers from it. There are a few caveats like [this](https://github.com/cockroachdb/cockroach/blob/8e6e4090457565a41bc3bd8ea954e437030d1c49/pkg/raft/raft.go#L1153-L1179) uncontrolled scan happening in campaigning steps. But it can be optimized out (will file an issue).